### ISSUE TITLE
[RFC] use `shfmt` for all hack shell scripts?

### ISSUE THREAD

**Issue by AkihiroSuda** [moby/moby#38055]
https://github.com/mvdan/sh seems useful for keeping indent style consistent across all shell scripts

```console
$ git rev-parse HEAD
503b1a9b6f24488db6a67f7ba24258e4ff5ea2a7
$ git grep --name-only -e '^#!' | xargs shfmt -l -w
$ git diff --shortstat 
 55 files changed, 938 insertions(+), 867 deletions(-)
```

@thaJeztah @vdemeester WDYT?

**Comment by AkihiroSuda** [comment]
cc @tamalsaha (thanks for the info!)

**Comment by vdemeester** [comment]
cc @tianon 
But I'm :100: on a `fmt` binary for sh that we could use in `validate` :angel: :pray:

**Comment by AkihiroSuda** [comment]
branch with shfmt is here: https://github.com/AkihiroSuda/docker/commits/shfmt

**Comment by thaJeztah** [comment]
❤️ 

@tianon @albers ptal

**Comment by thaJeztah** [comment]
oh! sorry for the double ping @tianon 😊

**Comment by AkihiroSuda** [comment]
Another candidate is https://www.shellcheck.net 

I'm not sure which is better.

cc @tiborvass @jessfraz

**Comment by unclejack** [comment]
shfmt was OK the last time I've used it. I'd recommend it.

**Comment by mvdan** [comment]
Happy to help if you run into any complications :)

You can think of `shfmt` like `gofmt` (creative name), and `shellcheck` much like `golint` or `go vet`.

**Comment by thaJeztah** [comment]
I'm not familiar with any of these, so can't really make recommendations for one or the other. Just adding my 👍 for adding something, as it's both very easy to miss formatting issues during review (tabs/spaces), and easy to introduce these issues when editing files if your editor uses different settings.

**Comment by stevvooe** [comment]
I’d recommend both shfmt and shellcheck. Shellcheck does a great job of catching quoting bugs.

**Comment by tianon** [comment]
In my experience `shellcheck` causes a lot more headaches than the help it provides, but I've just taken a first look at `shfmt` (I'm unfamiliar with it) and it seems pretty reasonable (I tried it on a couple scripts and it provided reasonable suggestions, even in the face of odd features like `<<-EOF` -- and it uses the POSIX standard indentation and applies that appropriately to things like `<<-EOF`, so that's a pretty big attention to detail++ as far as I'm concerned).

**Comment by albers** [comment]
We already use `shellcheck` for Docker's bash completion. 

It works well but seems to be tailored for basic usage. In more advanced scripts that do things like assebling code in other languages, it gives a lot of false positives. For example, it complains that "Expressions don't expand in single quotes, use double quotes for that." where we construct `awk` command lines (that contain awk variables like `$1` inside single quotes), so we had to [disable several checks](https://github.com/docker/cli/blob/master/contrib/completion/bash/docker#L2).
But I can still recommend shellcheck.

For `shfmt`, I'll have to take a closer look.

**Comment by albers** [comment]
I skimmed over https://github.com/AkihiroSuda/docker/commit/b7aacb143706ac282de23cd8f7acb3de5113e690 and think the changes are sane.

One thing: in https://github.com/AkihiroSuda/docker/commit/b7aacb143706ac282de23cd8f7acb3de5113e690#diff-c63e117fda579aecfe0fb4b7cb35075eR124, indentation of a JSON document written by the script is modified. This might affect consumers.

Personally, I think there are cases where deviation from a strict formatting scheme increases readibility,  see https://github.com/AkihiroSuda/docker/commit/b7aacb143706ac282de23cd8f7acb3de5113e690#diff-c63e117fda579aecfe0fb4b7cb35075eL52 for an example. But this is not a big deal for me.

Summing up this and my last comment, I'm :+1: for using `shfmt` and `shellcheck` here and in the CLI's bash completion.

**Comment by AkihiroSuda** [comment]
@mvdan Could you take a look at the JSON issue which @albers mentioned?

**Comment by mvdan** [comment]
Sure.

> indentation of a JSON document written by the script is modified. This might affect consumers.

`<<-` documents are indented automatically. This is because the tabs are stripped from the input itself:

```
If  the  redirection operator is <<-, then all leading tab characters are stripped from input lines and the line containing delimiter.  This allows here-documents within shell scripts to be indented in a natural fashion.
```

If you want the indentation to stay exactly as it is, use `<<`. The only other option is to somehow teach `shfmt` about custom indentation in `<<-` heredocs. I can't think of a simple set of rules that would make that possible, besides never changing the indentation.

> Indentation should not be changed here. (the comments)

Indeed, that's a bug. Will investigate.

> Here, I prefer the two-command style because it expresses the rhythm of the code more clearly.

https://github.com/mvdan/sh/issues/261 would likely fix this. That is, the option to loosen the "maximum one command per line" rule.

**Comment by albers** [comment]
@mvdan A better treatment of whitespace in `<<-` might be to determine the indentation from the first line of the here document and only strip that amount of whitespace from the following lines.

**Comment by mvdan** [comment]
What I quoted earlier is from `man bash` - I am not in control of how `<<-` works when run.

Unless you mean that we should only change the indentation relative to the first line in the `<<-` body? That is, lines that are more indented relative to it, would remain that way.

**Comment by albers** [comment]
I was referring to source code formatting. I'd like to preserve indentation below the top level indentation inside here documents.
In https://github.com/AkihiroSuda/docker/commit/b7aacb143706ac282de23cd8f7acb3de5113e690#diff-c63e117fda579aecfe0fb4b7cb35075eR124, the json hierarchy is entirely flattened. I would prefer shfmt not to change code layout in this particular example.

You are right, both code formatting variants will produce the same output at runtime (because tabs are used for indentation). I was wrong in my comment above, this code change will not affect behaviour of the script.

**Comment by mvdan** [comment]
Makes sense - I've filed an issue about it. Shouldn't be hard to implement, although I likely won't do a release for another couple of weeks, as I'm in the middle of some refactors. If you can live with the current syntax for now, I'd leave a TODO and improve the format once there's a new release.

**Comment by albers** [comment]
> If you can live with the current syntax for now, I'd leave a TODO and improve the format once there's a new release.

This is definitely not a blocker.

**Comment by albers** [comment]
@mvdan Formatting of the two sections is now perfect. Thanks a lot for the quick response.

I also compared the new formatting results in the *./contrib* dir with those in @AkihiroSuda's branch in search for side effects.

The only additional change I found was in `contrib/report-issue.sh`  where `template()` is now formatted differently in the "empty" lines (one tab less). This is probably how it should work and does not change the output of the function.

**Comment by AkihiroSuda** [comment]
PR: https://github.com/moby/moby/pull/40607
